File: propagator_simulation.m
%
% Propagator Simulation
%
% Copyright (c) 2015 WatSat-ADCS
% Licensed under the MIT license.
%
% authors: Jason Pye (j2pye@uwaterloo.ca)
%
% Change log:
% 2015-10-02 (JP) - Initial release
%


%% Next version: make functions for the python calls!

clear all
close all

% Initialise
system("python -c 'from clock import resetClock; resetClock()'");
system("python -c 'from satTLE import resetTLE; resetTLE()'");


% Time
% Gregorian calendar starts at: year=-4713, month=11, day=24, hour=12, minute=0, second=0

[status, result] = system("python -c 'from clock import greg2jdate; print greg2jdate(-4713,11,24,12 + 5,0,0)'"); % year, month, day, hour, minute, second
time_total = str2num(result);

[status, result] = system("python -c 'from clock import greg2jdate; print greg2jdate(-4713,11,24,12,0 + 5,0)'");
time_interval = str2num(result);

[status, result] = system("python -c 'from clock import readClock; print readClock()'");
time_var = str2num(result);


% Propagate

[status,result] = system("python -c 'from satTLE import currentOrbitState; (px,py,pz),(vx,vy,vz) = currentOrbitState(); print px; print py; print pz; print vx; print vy; print vz;'");
num_result = str2num(result);
position = num_result(1:3);
velocity = num_result(4:6);

timeIx = 1;
while ( time_var(timeIx) + time_interval < time_var(1) + time_total )

  system(["python -c 'from clock import tickClock; tickClock(",num2str(time_interval),")'"]);

  [status, result] = system("python -c 'from clock import readClock; print readClock()'");
  time_var = [time_var, str2num(result)];

  [status,result] = system("python -c 'from satTLE import currentOrbitState; (px,py,pz),(vx,vy,vz) = currentOrbitState(); print px; print py; print pz; print vx; print vy; print vz;'");
  num_result = str2num(result);
  position = [position, num_result(1:3)];
  velocity = [velocity, num_result(4:6)];

  timeIx = timeIx + 1;

end



% Plot

figure
hold on

subplot(3,2,1)
% plot(time_var, position(1,:), 'b-', 'Linewidth',2)
plot(position(1,:), 'b-', 'Linewidth',2)
ylabel('Position, x (km)','FontSize',16)
xlabel('Julian Date (not yet)','FontSize',16)
set(gca,'FontSize',16)

subplot(3,2,3)
plot(position(2,:), 'b-', 'Linewidth',2)
ylabel('Position, y (km)','FontSize',16)
xlabel('Julian Date (not yet)','FontSize',16)
set(gca,'FontSize',16)

subplot(3,2,5)
plot(position(3,:), 'b-', 'Linewidth',2)
ylabel('Position, z (km)','FontSize',16)
xlabel('Julian Date (not yet)','FontSize',16)
set(gca,'FontSize',16)

subplot(3,2,2)
plot(velocity(1,:), 'b-', 'Linewidth',2)
ylabel('Velocity, x (km/s .. I think)','FontSize',16)
xlabel('Julian Date (not yet)','FontSize',16)
set(gca,'FontSize',16)

subplot(3,2,4)
plot(velocity(2,:), 'b-', 'Linewidth',2)
ylabel('Velocity, y (km/s .. I think)','FontSize',16)
xlabel('Julian Date (not yet)','FontSize',16)
set(gca,'FontSize',16)

subplot(3,2,6)
plot(velocity(3,:), 'b-', 'Linewidth',2)
ylabel('Velocity, z (km/s .. I think)','FontSize',16)
xlabel('Julian Date (not yet)','FontSize',16)
set(gca,'FontSize',16)


figure
hold on

subplot(1,2,1)
% plot(time_var, position(1,:), 'b-', 'Linewidth',2)
plot( sqrt(position(1,:).^2 + position(2,:).^2 + position(3,:).^2 ), 'b-', 'Linewidth',2)
ylabel('Distance (km)','FontSize',16)
xlabel('Julian Date (not yet)','FontSize',16)
set(gca,'FontSize',16)

subplot(1,2,2)
plot( sqrt(velocity(1,:).^2 + velocity(2,:).^2 + velocity(3,:).^2 ), 'b-', 'Linewidth',2)
ylabel('Speed (km/s .. I think)','FontSize',16)
xlabel('Julian Date (not yet)','FontSize',16)
set(gca,'FontSize',16)


figure

for timeIx = 1:numel(time_var)

  scatter3(position(1,1:timeIx),position(2,1:timeIx),position(3,1:timeIx),'filled')
  title('Position (km)','FontSize',16)
  xlim([-1e4,1e4])
  ylim([-1e4,1e4])
  zlim([-1e4,1e4])
  set(gca,'FontSize',14)
  pause(0.001)

end

File: propagator_simulation.c
/*
  Orbit Propagator Simulation

  Copyright (c) 2015 WatSat-ADCS
  Licensed under the MIT license.

  authors: Jason Pye (j2pye@uwaterloo.ca)

  Change log:
  2015-12-04 (JP) - Initial release
*/

#include <stdio.h>

#include "satPosition.h"
#include "clock.h"

#include "sgp4ext.h" // For invjday(...)


const char clock_filename[] = "dat/jdate.txt";
const char inittime_filename[] = "dat/init_time.txt";
const char rv_filename[] = "dat/rv_est.txt";
const char initTLE_filename[] = "dat/tle_init.txt";

const char csvout_filename[] = "simulation_validation/simulation_data.csv";


int main() {

  // Initialise
  resetClock(clock_filename, inittime_filename);
  resetRV(rv_filename, initTLE_filename);

  // Prepare output file
  FILE *csvptr;
  csvptr = fopen( csvout_filename, "w" );
  if ( csvptr == NULL ) {
    printf("There was an error opening file: %s\n", csvout_filename);
    return -1;
  }

  fprintf(csvptr, "YEAR,MONTH,DAY,HOUR,MINUTE,SECOND,JULIAN_DATE,POSN_X,POSN_Y,POSN_Z,VEL_X,VEL_Y,VEL_Z\n");


  // Simulation

  double init_time = readClock(clock_filename);
  double total_time = 0/1440.0;
//  double total_time = 1*(365); // One year in jdate
  double incr_time = 1/1440.0; // One minute in jdate

  double rvtime, posn[3], vel[3];
  int year, mon, day, hr, minute;
  double sec;

  while ( rvtime - init_time < total_time ) {

    // Gets current orbit state and stores in rvtime, posn, vel
    currentOrbitState(rv_filename, initTLE_filename, clock_filename, &rvtime, posn, vel);

    // Write data to output file
    invjday(rvtime, year, mon, day, hr, minute, sec);
    fprintf(csvptr, "%d,%d,%d,%d,%d,%.5lf", year, mon, day, hr, minute, sec);
    fprintf(csvptr, ",%.5lf", rvtime);
    int i;
    for (i = 0; i < 3; i++)
      fprintf(csvptr, ",%.5lf", posn[i]);
    for (i = 0; i < 3; i++)
      fprintf(csvptr, ",%.5lf", vel[i]);
    fprintf(csvptr, "\n");

    // Ticks clock
    tickClock(clock_filename, incr_time);

  }

  fclose(csvptr);

  return 0;
}
